monty_profiling/frameworks/models/evidence_matching.py

```diff
@@ -0,0 +1,316 @@
+        num_model_voxels_per_dim=50,
+        use_multithreading=True,
+        gsg_class=FlopCountingEvidenceGoalStateGenerator,
+        gsg_args=None,
+        *args,
+        **kwargs,
+    ):
+        # Call parent's __init__ with all parameters
+        super().__init__(
+            max_match_distance=max_match_distance,
+            tolerances=tolerances,
+            feature_weights=feature_weights,
+            feature_evidence_increment=feature_evidence_increment,
+            max_nneighbors=max_nneighbors,
+            initial_possible_poses=initial_possible_poses,
+            evidence_update_threshold=evidence_update_threshold,
+            vote_evidence_threshold=vote_evidence_threshold,
+            past_weight=past_weight,
+            present_weight=present_weight,
+            vote_weight=vote_weight,
+            object_evidence_threshold=object_evidence_threshold,
+            x_percent_threshold=x_percent_threshold,
+            path_similarity_threshold=path_similarity_threshold,
+            pose_similarity_threshold=pose_similarity_threshold,
+            required_symmetry_evidence=required_symmetry_evidence,
+            graph_delta_thresholds=graph_delta_thresholds,
+            max_graph_size=max_graph_size,
+            max_nodes_per_graph=max_nodes_per_graph,
+            num_model_voxels_per_dim=num_model_voxels_per_dim,
+            use_multithreading=use_multithreading,
+            gsg_class=gsg_class,
+            gsg_args=gsg_args,
+            *args,
+            **kwargs,
+        )
+        self.flop_counter = None
+
+    def _update_evidence_with_vote(self, state_votes, graph_id):
+        """Count FLOPs for KDTree operations during vote processing."""
+        # Extract vote information
+        graph_location_vote = np.zeros((len(state_votes), 3))
+        vote_evidences = np.zeros(len(state_votes))
+        for n, vote in enumerate(state_votes):
+            graph_location_vote[n] = vote.location
+            vote_evidences[n] = vote.confidence
+
+        # Execute KDTree query
+        vote_location_tree = KDTree(
+            graph_location_vote,
+            leafsize=40,
+        )
+        vote_nn = 3  # TODO: Make this a parameter?
+        if graph_location_vote.shape[0] < vote_nn:
+            vote_nn = graph_location_vote.shape[0]
+
+        # Account for FLOPs for KDTree Construction
+        num_reference_points = len(graph_location_vote)
+        dim1 = self.possible_locations[graph_id].shape[1]
+        kdtree_flops = int(num_reference_points * np.log2(num_reference_points) * dim1)
+
+        (radius_node_dists, radius_node_ids) = vote_location_tree.query(
+            self.possible_locations[graph_id],
+            k=vote_nn,
+            p=2,
+            workers=1,
+        )
+
+        # Account for FLOPs for KDTree query
+        num_search_points = len(self.possible_locations[graph_id])
+        num_reference_points = len(graph_location_vote)
+
+        # Tree Traversal FLOPs
```

Comment:
Similarly for this?

Reply:
Ok interesting thanks - by balanced this presumably means that points are fairly evenly distributed in space? I think that's a reasonable assumption given the structure of our objects, i.e. they don't tend to bunch up in just 2-3 clouds.